rfcs/2022-03-15-11851-ingest-opentelemtry-traces.md

```diff
@@ -0,0 +1,368 @@
+
+### User Experience
+
+- User would point OpenTelemtry tracing lib directly to a local Vector deployement
+- Vector would be configured with a minimal config looking like:
+
+```yaml
+sources:
+  otlp_traces:
+    type: opentelemetry_traces
+    address: "[::]:8081"
+    mode: grpc
+
+sinks:
+  dd_trace:
+    type: datadog_traces
+    default_api_key: 12345678abcdef
+    inputs:
+     - otlp_traces
+```
+
+And it should just work.
+
+### Implementation
+
+- `opentelemetry_traces` or simply `opentelemetry` sources:
+  - The gRPC variant would use Tonic to spawn a gRPC server (like the `vector` source in its v2 variation) and directly
+    use the [offical gRPC service definitions][otlp-grpc-def].
+  - HTTP variant would use a Warp server and attempt to decode protobuf payloads, as per the [specification][otlp-http],
+    payloads are encoded using protobuf either in binary format or in JSON format ([Protobuf schemas][otlp-proto-def]).
+    All the expected behaviours regarding the kind of requests/responses code and sequence are clearly defined as well
+    as the default URL path (`/v1/traces` for traces).
+- Internal traces representation/normalization, two options are opened see [outstanding
+  questions](#outstanding-questions), but the consensus is leaning towards a new dedicated container that would:
+  - Move away from from the current implementation that relies on `LogEvent` to a dedicated container, the
+    implementation will stay  in [./lib/vector-core/src/event/trace.rs][current-trace-in-vector]
+  - Borrow most of its semantic from the Opentelemetry Traces format, the `TraceEvent` would then evolve toward a
+    concrete dedicated container and contains a slice of spans, the span being directly inspired by the Opentelemtry
+    specification.
+  - Some top-level information (Like trace ID, trace-wide tags/metrics, the original format)
+  - Trace would not get native `VrlTarget` representation anymore, there is a bigger discussion there that should
+    probably be adressed separately. As an interim measure few fields may be exposed (like trace ID & trace-wide
+    tags).
+- APM stats computation:
+  - Implement a similar logic that the one done in the Datadog OTLP exporter, this would allow user to use multiple
+    Datadog product with Opentelemetry traces and get the same consistent behaviour in all circumstances. APM stats
+    computation is hooked [there][apm-stats-computation] in the Datadog exporter. But as this is go code it relies on
+    the [Agent codebase][agent-code-for-otlp-exporter] to do the [actual computation][agent-handle-span].
+  - Where the APM stats computations is still under discusion, see [outstanding questions](#outstanding-questions)
+
+## Rationale
+
+- Opentelemetry is the de-facto standard for traces, so supporting it at some point is mandatory. Note that this
+  consideration is wider than just traces as metrics (and logs) are addressed by the Opentelemetry project.
+
+## Drawbacks
+
+N/A
```

Comment:
Most of the traces format designed for remote submission & processing seems to be able to fit into an opentelemetry inspired container. There are some trace receiver alreadry explicitly doing that [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver) for AWS X-ray and Skywalking traces. Other trace formats that target remote processing like the [new relic format](https://docs.newrelic.com/docs/distributed-tracing/trace-api/report-new-relic-format-traces-trace-api/) seems to be suitable to fit inside an opentelemetry based container. However if we think about trace format like perf data or CTF, that are mostly processed locally, they are really different, and if we want to support that kind of traces this will probably required dedicated container, and this is indeed a drawback, I will update the RFC to reflect that.